File: Neetcode/Graphs/validtree.cpp
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

class Solution {
 public:
  unordered_set<int> visited;
  bool dfs(unordered_map<int, vector<int>> nes, int curr, int prev) {
    if (visited.find(curr) != visited.end()) {
      return false;
    }
    visited.insert(curr);

    for (auto ne : nes[curr]) {
      if (ne != prev) {
        if (!dfs(nes, ne, curr)) {
          return false;
        }
      }
    }
    return true;
  }

  bool validTree(int n, vector<vector<int>>& edges) {
    unordered_map<int, vector<int>> nes;
    for (auto edge : edges) {
      nes[edge[0]].push_back(edge[1]);
      nes[edge[1]].push_back(edge[0]);
    }
    return dfs(nes, 0, -1) && (n == visited.size());
  }
};

File: Neetcode/Graphs/validtree.py
# https://neetcode.io/problems/valid-tree
# https://leetcode.com/problems/graph-valid-tree/
# https://www.lintcode.com/problem/178/

from typing import List


class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if not n:
            return True

        adj = {i: [] for i in range(n)}
        for a, b in edges:
            adj[a].append(b)
            adj[b].append(a)

        visited = set()

        def dfs(i, prev):
            if i in visited:
                return False

            visited.add(i)
            for ne in adj[i]:
                if ne == prev:
                    continue
                if not dfs(ne, i):
                    return False
            return True

        return dfs(0, -1) and n == len(visited)
